File: src/src/main/resources/mmlspark/_SimpleHTTPTransformer.py
# Copyright (C) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See LICENSE in project root for information.


import sys
if sys.version >= '3':
    basestring = str

from pyspark.ml.param.shared import *
from pyspark import keyword_only
from pyspark.ml.util import JavaMLReadable, JavaMLWritable
from pyspark.ml.wrapper import JavaTransformer, JavaEstimator, JavaModel
from pyspark.ml.common import inherit_doc
from mmlspark.Utils import *
from mmlspark.TypeConversionUtils import generateTypeConverter, complexTypeConverter

@inherit_doc
class _SimpleHTTPTransformer(ComplexParamsMixin, JavaMLReadable, JavaMLWritable, JavaTransformer):
    """


    Args:

        concurrency (int): max number of concurrent calls (default: 1)
        concurrentTimeout (double): max number seconds to wait on futures if concurrency >= 1 (default: 100.0)
        errorCol (str): column to hold http errors (default: [self.uid]_errors)
        flattenOutputBatches (bool): whether to flatten the output batches
        handler (object): Which strategy to use when handling requests (default: UserDefinedFunction(<function2>,StringType,None))
        inputCol (str): The name of the input column
        inputParser (object): format to parse the column to (default: JSONInputParser_1e823ee240ad)
        miniBatcher (object): Minibatcher to use
        outputCol (str): The name of the output column
        outputParser (object): format to parse the column to
        timeout (double): number of seconds to wait before closing the connection (default: 60.0)
    """

    @keyword_only
    def __init__(self, concurrency=1, concurrentTimeout=100.0, errorCol=None, flattenOutputBatches=None, handler=None, inputCol=None, inputParser=None, miniBatcher=None, outputCol=None, outputParser=None, timeout=60.0):
        super(_SimpleHTTPTransformer, self).__init__()
        self._java_obj = self._new_java_obj("com.microsoft.ml.spark.SimpleHTTPTransformer")
        self._cache = {}
        self.concurrency = Param(self, "concurrency", "concurrency: max number of concurrent calls (default: 1)")
        self._setDefault(concurrency=1)
        self.concurrentTimeout = Param(self, "concurrentTimeout", "concurrentTimeout: max number seconds to wait on futures if concurrency >= 1 (default: 100.0)")
        self._setDefault(concurrentTimeout=100.0)
        self.errorCol = Param(self, "errorCol", "errorCol: column to hold http errors (default: [self.uid]_errors)")
        self._setDefault(errorCol=self.uid + "_errors")
        self.flattenOutputBatches = Param(self, "flattenOutputBatches", "flattenOutputBatches: whether to flatten the output batches")
        self.handler = Param(self, "handler", "handler: Which strategy to use when handling requests (default: UserDefinedFunction(<function2>,StringType,None))", generateTypeConverter("handler", self._cache, complexTypeConverter))
        self.inputCol = Param(self, "inputCol", "inputCol: The name of the input column")
        self.inputParser = Param(self, "inputParser", "inputParser: format to parse the column to (default: JSONInputParser_1e823ee240ad)", generateTypeConverter("inputParser", self._cache, complexTypeConverter))
        self.miniBatcher = Param(self, "miniBatcher", "miniBatcher: Minibatcher to use", generateTypeConverter("miniBatcher", self._cache, complexTypeConverter))
        self.outputCol = Param(self, "outputCol", "outputCol: The name of the output column")
        self.outputParser = Param(self, "outputParser", "outputParser: format to parse the column to", generateTypeConverter("outputParser", self._cache, complexTypeConverter))
        self.timeout = Param(self, "timeout", "timeout: number of seconds to wait before closing the connection (default: 60.0)")
        self._setDefault(timeout=60.0)
        if hasattr(self, "_input_kwargs"):
            kwargs = self._input_kwargs
        else:
            kwargs = self.__init__._input_kwargs
        self.setParams(**kwargs)

    @keyword_only
    def setParams(self, concurrency=1, concurrentTimeout=100.0, errorCol=None, flattenOutputBatches=None, handler=None, inputCol=None, inputParser=None, miniBatcher=None, outputCol=None, outputParser=None, timeout=60.0):
        """
        Set the (keyword only) parameters

        Args:

            concurrency (int): max number of concurrent calls (default: 1)
            concurrentTimeout (double): max number seconds to wait on futures if concurrency >= 1 (default: 100.0)
            errorCol (str): column to hold http errors (default: [self.uid]_errors)
            flattenOutputBatches (bool): whether to flatten the output batches
            handler (object): Which strategy to use when handling requests (default: UserDefinedFunction(<function2>,StringType,None))
            inputCol (str): The name of the input column
            inputParser (object): format to parse the column to (default: JSONInputParser_1e823ee240ad)
            miniBatcher (object): Minibatcher to use
            outputCol (str): The name of the output column
            outputParser (object): format to parse the column to
            timeout (double): number of seconds to wait before closing the connection (default: 60.0)
        """
        if hasattr(self, "_input_kwargs"):
            kwargs = self._input_kwargs
        else:
            kwargs = self.__init__._input_kwargs
        return self._set(**kwargs)

    def setConcurrency(self, value):
        """

        Args:

            concurrency (int): max number of concurrent calls (default: 1)

        """
        self._set(concurrency=value)
        return self


    def getConcurrency(self):
        """

        Returns:

            int: max number of concurrent calls (default: 1)
        """
        return self.getOrDefault(self.concurrency)


    def setConcurrentTimeout(self, value):
        """

        Args:

            concurrentTimeout (double): max number seconds to wait on futures if concurrency >= 1 (default: 100.0)

        """
        self._set(concurrentTimeout=value)
        return self


    def getConcurrentTimeout(self):
        """

        Returns:

            double: max number seconds to wait on futures if concurrency >= 1 (default: 100.0)
        """
        return self.getOrDefault(self.concurrentTimeout)


    def setErrorCol(self, value):
        """

        Args:

            errorCol (str): column to hold http errors (default: [self.uid]_errors)

        """
        self._set(errorCol=value)
        return self


    def getErrorCol(self):
        """

        Returns:

            str: column to hold http errors (default: [self.uid]_errors)
        """
        return self.getOrDefault(self.errorCol)


    def setFlattenOutputBatches(self, value):
        """

        Args:

            flattenOutputBatches (bool): whether to flatten the output batches

        """
        self._set(flattenOutputBatches=value)
        return self


    def getFlattenOutputBatches(self):
        """

        Returns:

            bool: whether to flatten the output batches
        """
        return self.getOrDefault(self.flattenOutputBatches)


    def setHandler(self, value):
        """

        Args:

            handler (object): Which strategy to use when handling requests (default: UserDefinedFunction(<function2>,StringType,None))

        """
        self._set(handler=value)
        return self


    def getHandler(self):
        """

        Returns:

            object: Which strategy to use when handling requests (default: UserDefinedFunction(<function2>,StringType,None))
        """
        return self._cache.get("handler", None)


    def setInputCol(self, value):
        """

        Args:

            inputCol (str): The name of the input column

        """
        self._set(inputCol=value)
        return self


    def getInputCol(self):
        """

        Returns:

            str: The name of the input column
        """
        return self.getOrDefault(self.inputCol)


    def setInputParser(self, value):
        """

        Args:

            inputParser (object): format to parse the column to (default: JSONInputParser_1e823ee240ad)

        """
        self._set(inputParser=value)
        return self


    def getInputParser(self):
        """

        Returns:

            object: format to parse the column to (default: JSONInputParser_1e823ee240ad)
        """
        return self._cache.get("inputParser", None)


    def setMiniBatcher(self, value):
        """

        Args:

            miniBatcher (object): Minibatcher to use

        """
        self._set(miniBatcher=value)
        return self


    def getMiniBatcher(self):
        """

        Returns:

            object: Minibatcher to use
        """
        return self._cache.get("miniBatcher", None)


    def setOutputCol(self, value):
        """

        Args:

            outputCol (str): The name of the output column

        """
        self._set(outputCol=value)
        return self


    def getOutputCol(self):
        """

        Returns:

            str: The name of the output column
        """
        return self.getOrDefault(self.outputCol)


    def setOutputParser(self, value):
        """

        Args:

            outputParser (object): format to parse the column to

        """
        self._set(outputParser=value)
        return self


    def getOutputParser(self):
        """

        Returns:

            object: format to parse the column to
        """
        return self._cache.get("outputParser", None)


    def setTimeout(self, value):
        """

        Args:

            timeout (double): number of seconds to wait before closing the connection (default: 60.0)

        """
        self._set(timeout=value)
        return self


    def getTimeout(self):
        """

        Returns:

            double: number of seconds to wait before closing the connection (default: 60.0)
        """
        return self.getOrDefault(self.timeout)



    @classmethod
    def read(cls):
        """ Returns an MLReader instance for this class. """
        return JavaMMLReader(cls)

    @staticmethod
    def getJavaPackage():
        """ Returns package name String. """
        return "com.microsoft.ml.spark.SimpleHTTPTransformer"

    @staticmethod
    def _from_java(java_stage):
        module_name=_SimpleHTTPTransformer.__module__
        module_name=module_name.rsplit(".", 1)[0] + ".SimpleHTTPTransformer"
        return from_java(java_stage, module_name)

File: dotnetSanityCheck/dotnetClasses/SimpleHTTPTransformer.cs
// Copyright (C) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE in project root for information.


using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using Microsoft.Spark.ML.Feature.Param;
using Microsoft.Spark.Interop;
using Microsoft.Spark.Interop.Ipc;
using Microsoft.Spark.ML.Feature;
using Microsoft.Spark.Sql;
using Microsoft.Spark.Sql.Types;
using Microsoft.Spark.Utils;

using SynapseML.Dotnet.Utils;


namespace Synapse.ML.Io.Http
{
    /// <summary>
    /// <see cref="SimpleHTTPTransformer"/> implements SimpleHTTPTransformer
    /// </summary>
    public class SimpleHTTPTransformer : JavaTransformer, IJavaMLWritable, IJavaMLReadable<SimpleHTTPTransformer>
    {
        private static readonly string s_className = "com.microsoft.azure.synapse.ml.io.http.SimpleHTTPTransformer";

        /// <summary>
        /// Creates a <see cref="SimpleHTTPTransformer"/> without any parameters.
        /// </summary>
        public SimpleHTTPTransformer() : base(s_className)
        {
        }

        /// <summary>
        /// Creates a <see cref="SimpleHTTPTransformer"/> with a UID that is used to give the
        /// <see cref="SimpleHTTPTransformer"/> a unique ID.
        /// </summary>
        /// <param name="uid">An immutable unique ID for the object and its derivatives.</param>
        public SimpleHTTPTransformer(string uid) : base(s_className, uid)
        {
        }

        internal SimpleHTTPTransformer(JvmObjectReference jvmObject) : base(jvmObject)
        {
        }

        /// <summary>
        /// Sets concurrency value for <see cref="concurrency"/>
        /// </summary>
        /// <param name="concurrency">
        /// max number of concurrent calls
        /// </param>
        /// <returns> New SimpleHTTPTransformer object </returns>
        public SimpleHTTPTransformer SetConcurrency(int value) =>
            WrapAsSimpleHTTPTransformer(Reference.Invoke("setConcurrency", (object)value));
        
        /// <summary>
        /// Sets concurrentTimeout value for <see cref="concurrentTimeout"/>
        /// </summary>
        /// <param name="concurrentTimeout">
        /// max number seconds to wait on futures if concurrency >= 1
        /// </param>
        /// <returns> New SimpleHTTPTransformer object </returns>
        public SimpleHTTPTransformer SetConcurrentTimeout(double value) =>
            WrapAsSimpleHTTPTransformer(Reference.Invoke("setConcurrentTimeout", (object)value));
        
        /// <summary>
        /// Sets errorCol value for <see cref="errorCol"/>
        /// </summary>
        /// <param name="errorCol">
        /// column to hold http errors
        /// </param>
        /// <returns> New SimpleHTTPTransformer object </returns>
        public SimpleHTTPTransformer SetErrorCol(string value) =>
            WrapAsSimpleHTTPTransformer(Reference.Invoke("setErrorCol", (object)value));
        
        /// <summary>
        /// Sets flattenOutputBatches value for <see cref="flattenOutputBatches"/>
        /// </summary>
        /// <param name="flattenOutputBatches">
        /// whether to flatten the output batches
        /// </param>
        /// <returns> New SimpleHTTPTransformer object </returns>
        public SimpleHTTPTransformer SetFlattenOutputBatches(bool value) =>
            WrapAsSimpleHTTPTransformer(Reference.Invoke("setFlattenOutputBatches", (object)value));
        
        /// <summary>
        /// Sets handler value for <see cref="handler"/>
        /// </summary>
        /// <param name="handler">
        /// Which strategy to use when handling requests
        /// </param>
        /// <returns> New SimpleHTTPTransformer object </returns>
        public SimpleHTTPTransformer SetHandler(object value) =>
            WrapAsSimpleHTTPTransformer(Reference.Invoke("setHandler", value));
        
        /// <summary>
        /// Sets inputCol value for <see cref="inputCol"/>
        /// </summary>
        /// <param name="inputCol">
        /// The name of the input column
        /// </param>
        /// <returns> New SimpleHTTPTransformer object </returns>
        public SimpleHTTPTransformer SetInputCol(string value) =>
            WrapAsSimpleHTTPTransformer(Reference.Invoke("setInputCol", (object)value));
        
        /// <summary>
        /// Sets inputParser value for <see cref="inputParser"/>
        /// </summary>
        /// <param name="inputParser">
        /// format to parse the column to
        /// </param>
        /// <returns> New SimpleHTTPTransformer object </returns>
        public SimpleHTTPTransformer SetInputParser(JavaTransformer value) =>
            WrapAsSimpleHTTPTransformer(Reference.Invoke("setInputParser", (object)value));
        
        /// <summary>
        /// Sets miniBatcher value for <see cref="miniBatcher"/>
        /// </summary>
        /// <param name="miniBatcher">
        /// Minibatcher to use
        /// </param>
        /// <returns> New SimpleHTTPTransformer object </returns>
        public SimpleHTTPTransformer SetMiniBatcher(JavaTransformer value) =>
            WrapAsSimpleHTTPTransformer(Reference.Invoke("setMiniBatcher", (object)value));
        
        /// <summary>
        /// Sets outputCol value for <see cref="outputCol"/>
        /// </summary>
        /// <param name="outputCol">
        /// The name of the output column
        /// </param>
        /// <returns> New SimpleHTTPTransformer object </returns>
        public SimpleHTTPTransformer SetOutputCol(string value) =>
            WrapAsSimpleHTTPTransformer(Reference.Invoke("setOutputCol", (object)value));
        
        /// <summary>
        /// Sets outputParser value for <see cref="outputParser"/>
        /// </summary>
        /// <param name="outputParser">
        /// format to parse the column to
        /// </param>
        /// <returns> New SimpleHTTPTransformer object </returns>
        public SimpleHTTPTransformer SetOutputParser(JavaTransformer value) =>
            WrapAsSimpleHTTPTransformer(Reference.Invoke("setOutputParser", (object)value));
        
        /// <summary>
        /// Sets timeout value for <see cref="timeout"/>
        /// </summary>
        /// <param name="timeout">
        /// number of seconds to wait before closing the connection
        /// </param>
        /// <returns> New SimpleHTTPTransformer object </returns>
        public SimpleHTTPTransformer SetTimeout(double value) =>
            WrapAsSimpleHTTPTransformer(Reference.Invoke("setTimeout", (object)value));

        
        /// <summary>
        /// Gets concurrency value for <see cref="concurrency"/>
        /// </summary>
        /// <returns>
        /// concurrency: max number of concurrent calls
        /// </returns>
        public int GetConcurrency() =>
            (int)Reference.Invoke("getConcurrency");
        
        
        /// <summary>
        /// Gets concurrentTimeout value for <see cref="concurrentTimeout"/>
        /// </summary>
        /// <returns>
        /// concurrentTimeout: max number seconds to wait on futures if concurrency >= 1
        /// </returns>
        public double GetConcurrentTimeout() =>
            (double)Reference.Invoke("getConcurrentTimeout");
        
        
        /// <summary>
        /// Gets errorCol value for <see cref="errorCol"/>
        /// </summary>
        /// <returns>
        /// errorCol: column to hold http errors
        /// </returns>
        public string GetErrorCol() =>
            (string)Reference.Invoke("getErrorCol");
        
        
        /// <summary>
        /// Gets flattenOutputBatches value for <see cref="flattenOutputBatches"/>
        /// </summary>
        /// <returns>
        /// flattenOutputBatches: whether to flatten the output batches
        /// </returns>
        public bool GetFlattenOutputBatches() =>
            (bool)Reference.Invoke("getFlattenOutputBatches");
        
        
        /// <summary>
        /// Gets handler value for <see cref="handler"/>
        /// </summary>
        /// <returns>
        /// handler: Which strategy to use when handling requests
        /// </returns>
        public object GetHandler() => Reference.Invoke("getHandler");
        
        
        /// <summary>
        /// Gets inputCol value for <see cref="inputCol"/>
        /// </summary>
        /// <returns>
        /// inputCol: The name of the input column
        /// </returns>
        public string GetInputCol() =>
            (string)Reference.Invoke("getInputCol");
        
        
        /// <summary>
        /// Gets inputParser value for <see cref="inputParser"/>
        /// </summary>
        /// <returns>
        /// inputParser: format to parse the column to
        /// </returns>
        public JavaTransformer GetInputParser()
        {
            JvmObjectReference jvmObject = (JvmObjectReference)Reference.Invoke("getInputParser");
            Dictionary<string, Type> classMapping = JvmObjectUtils.ConstructJavaClassMapping(
                typeof(JavaTransformer),
                "s_className");
            JvmObjectUtils.TryConstructInstanceFromJvmObject(
                jvmObject,
                classMapping,
                out JavaTransformer instance);
            return instance;
        }
        
        
        /// <summary>
        /// Gets miniBatcher value for <see cref="miniBatcher"/>
        /// </summary>
        /// <returns>
        /// miniBatcher: Minibatcher to use
        /// </returns>
        public JavaTransformer GetMiniBatcher()
        {
            JvmObjectReference jvmObject = (JvmObjectReference)Reference.Invoke("getMiniBatcher");
            Dictionary<string, Type> classMapping = JvmObjectUtils.ConstructJavaClassMapping(
                typeof(JavaTransformer),
                "s_className");
            JvmObjectUtils.TryConstructInstanceFromJvmObject(
                jvmObject,
                classMapping,
                out JavaTransformer instance);
            return instance;
        }
        
        
        /// <summary>
        /// Gets outputCol value for <see cref="outputCol"/>
        /// </summary>
        /// <returns>
        /// outputCol: The name of the output column
        /// </returns>
        public string GetOutputCol() =>
            (string)Reference.Invoke("getOutputCol");
        
        
        /// <summary>
        /// Gets outputParser value for <see cref="outputParser"/>
        /// </summary>
        /// <returns>
        /// outputParser: format to parse the column to
        /// </returns>
        public JavaTransformer GetOutputParser()
        {
            JvmObjectReference jvmObject = (JvmObjectReference)Reference.Invoke("getOutputParser");
            Dictionary<string, Type> classMapping = JvmObjectUtils.ConstructJavaClassMapping(
                typeof(JavaTransformer),
                "s_className");
            JvmObjectUtils.TryConstructInstanceFromJvmObject(
                jvmObject,
                classMapping,
                out JavaTransformer instance);
            return instance;
        }
        
        
        /// <summary>
        /// Gets timeout value for <see cref="timeout"/>
        /// </summary>
        /// <returns>
        /// timeout: number of seconds to wait before closing the connection
        /// </returns>
        public double GetTimeout() =>
            (double)Reference.Invoke("getTimeout");

        
        /// <summary>
        /// Loads the <see cref="SimpleHTTPTransformer"/> that was previously saved using Save(string).
        /// </summary>
        /// <param name="path">The path the previous <see cref="SimpleHTTPTransformer"/> was saved to</param>
        /// <returns>New <see cref="SimpleHTTPTransformer"/> object, loaded from path.</returns>
        public static SimpleHTTPTransformer Load(string path) => WrapAsSimpleHTTPTransformer(
            SparkEnvironment.JvmBridge.CallStaticJavaMethod(s_className, "load", path));
        
        /// <summary>
        /// Saves the object so that it can be loaded later using Load. Note that these objects
        /// can be shared with Scala by Loading or Saving in Scala.
        /// </summary>
        /// <param name="path">The path to save the object to</param>
        public void Save(string path) => Reference.Invoke("save", path);
        
        /// <returns>a <see cref="JavaMLWriter"/> instance for this ML instance.</returns>
        public JavaMLWriter Write() =>
            new JavaMLWriter((JvmObjectReference)Reference.Invoke("write"));
        
        /// <returns>an <see cref="JavaMLReader"/> instance for this ML instance.</returns>
        public JavaMLReader<SimpleHTTPTransformer> Read() =>
            new JavaMLReader<SimpleHTTPTransformer>((JvmObjectReference)Reference.Invoke("read"));

        private static SimpleHTTPTransformer WrapAsSimpleHTTPTransformer(object obj) =>
            new SimpleHTTPTransformer((JvmObjectReference)obj);

        
    }
}

        